wagtail/contrib/redirects/signal_handlers.py

```diff
@@ -0,0 +1,94 @@
+import logging
+
+from typing import Optional
+
+from django.conf import settings
+from django.db.models import Q
+
+from wagtail.core.models import BaseLogEntry, Page, PageBase
+from wagtail.core.utils import BatchCreator, get_dummy_request
+
+from .models import Redirect
+
+
+logger = logging.getLogger(__name__)
+
+
+class BatchRedirectCreator(BatchCreator):
+    """
+    A specialized ``BatchCreator`` class for saving ``Redirect`` objects.
+    """
+
+    model = Redirect
+
+    def initialize_instance(self, kwargs):
+        # Automatically normalise paths when creating instances from kwargs
+        kwargs["old_path"] = Redirect.normalise_path(kwargs["old_path"])
+        return super().initialize_instance(kwargs)
+
+    def pre_process(self):
+        # delete any existing automatically-created redirects that might clash
+        # with the items in `self.items`
+        clashes_q = Q()
+        for item in self.items:
+            clashes_q |= Q(old_path=item.old_path, site_id=item.site_id)
+        Redirect.objects.filter(automatically_created=True).filter(clashes_q).delete()
+
+
+def autocreate_redirects(
+    sender: PageBase,
+    instance: Page,
+    url_path_before: str,
+    url_path_after: str,
+    log_entry: Optional[BaseLogEntry] = None,
+    **kwargs,
+) -> None:
+    if (
+        not getattr(settings, "WAGTAILREDIRECTS_AUTOCREATE", True)
+        or instance.is_site_root()
+    ):
+        return None
+
+    logger.info(f"Creating redirects for page: '{instance}' id={instance.id}")
+    batch = BatchRedirectCreator(max_size=2000, ignore_conflicts=True)
+    request = get_dummy_request()
```

Comment:
Would it be possible to implement this optimisation without creating a dummy request? Looks like we only need an object to hold the site route paths so they can be reused between calls, so there must be a simpler way to do that.

Also, the redirect object is passed in to `Site.find_for_request` but the dummy request would always have the same domain regardless of the page. Would this work when generating redirects on non-default sites?

Do you think it would be possible to pass the site to ``Page.get_url_paths`` instead of the request, and bypass the logic that finds the site? (ping @gasman)

Reply:
I see what you mean. Truth is, I find the various URL methods a little confusing to this day. I was under the impression that `request` could always be any old thing, but that isn't true for all of them. get_url_parts() does not care about the value type, but passing it a value called `request` that isn't a HttpRequest just to benefit from caching feels slightly more hacky to me. It's a strange API really. I would much rather url methods always receive a `HttpRequest`, and there be a lightweight helper (like this) to generate dummy ones when a real one isn't available. Or maybe a better route to consistency is to revise the URL methods so that `request` can *always* be any old value?

Reply:
>  get_url_parts() does NOT care about the value type

Actually, I have this wrong. Unless `request` is `None` or has both `get_host()` and `get_port()` methods, things break just beyond [this point](https://github.com/wagtail/wagtail/blob/main/wagtail/core/models/__init__.py#L1171) - so, using a `HttpRequest` is really the best option as far as I can see.

Reply:
> Do you think it would be possible to pass the site to Page.get_url_paths

Ah, you mean `get_url_parts` here.. the typo threw me. I'm with you now.

Reply:
> The dummy request would always have the same domain regardless of the page. Would this work when generating redirects on non-default sites?

Provided the page only belongs to a single site, it should be fine. I suppose we could instead determine which sites the page belongs to in advance, then repeatedly call `get_url_parts` for each site.

Reply:
Am open to suggestions on the name, but I've added [`Page.get_url_parts_for_all_sites()`](https://github.com/wagtail/wagtail/blob/774b82f97feacd3d7c2f635b7328b5b3b84aa3d8/wagtail/core/models/__init__.py#L1310-L1325) to help with this.

This and [`Page.get_relevant_site_root_paths`](https://github.com/wagtail/wagtail/blob/774b82f97feacd3d7c2f635b7328b5b3b84aa3d8/wagtail/core/models/__init__.py#L1190-L1204)  allow us to get the values we need without changing the signature or `get_url_parts()` and without using the dummy request.

Reply:
I realise now that a method separate to `get_url_parts()` will not work, as we need to pick up existing overrides for that method. So, I have added a `site` kwarg to `get_url_parts()` after all. The method was getting rather unwieldily though, so have broken it up a little to help with readability